File: JAVA/Leetcode/RotateImage.java
/*
Problem Statement:
You are given an n x n 2D matrix representing an image, rotate the image by 90 degrees (clockwise).
You have to rotate the image in-place, which means you have to modify the input 2D matrix directly. 
DO NOT allocate another 2D matrix and do the rotation.

Constraints:
n == matrix.length == matrix[i].length
1 <= n <= 20
-1000 <= matrix[i][j] <= 1000
*/

import java.util.Scanner;

public class RotateImage {

    public static void rotate(int[][] matrix) {
        int n = matrix.length;
        
        // Step 1: Transpose the matrix (swap matrix[i][j] with matrix[j][i])
        for (int i = 0; i < n; i++) {
            // Start j from i to avoid swapping elements back to their original positions
            for (int j = i; j < n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
        
        // Step 2: Reverse each row horizontally
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n / 2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][n - 1 - j];
                matrix[i][n - 1 - j] = temp;
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        try {
            System.out.println("Enter the size of the matrix (n):");
            int n = Integer.parseInt(scanner.nextLine().trim());
            
            int[][] matrix = new int[n][n];
            System.out.println("Enter the matrix rows (numbers separated by spaces):");
            for (int i = 0; i < n; i++) {
                String[] rowStrs = scanner.nextLine().trim().split("\\s+");
                for (int j = 0; j < n; j++) {
                    matrix[i][j] = Integer.parseInt(rowStrs[j]);
                }
            }
            
            rotate(matrix);
            
            System.out.println("Rotated Matrix:");
            for (int i = 0; i < n; i++) {
                System.out.print("[");
                for (int j = 0; j < n; j++) {
                    System.out.print(matrix[i][j] + (j == n - 1 ? "" : ", "));
                }
                System.out.println("]");
            }
            
        } catch (Exception e) {
            System.out.println("Invalid input format detected. Please try again.");
        } finally {
            scanner.close();
        }
    }
}

File: PYTHON/RotateImage.py
'''
Problem Statement:
You are given an n x n 2D matrix representing an image, rotate the image by 90 degrees (clockwise).
You have to rotate the image in-place, which means you have to modify the input 2D matrix directly. 
DO NOT allocate another 2D matrix and do the rotation.

Constraints:
n == matrix.length == matrix[i].length
1 <= n <= 20
-1000 <= matrix[i][j] <= 1000
'''

def rotate(matrix: list[list[int]]) -> None:
    """
    Do not return anything, modify matrix in-place instead.
    """
    n = len(matrix)
    
    # Step 1: Transpose the matrix
    for i in range(n):
        for j in range(i, n):
            # Python's tuple unpacking allows us to swap without a temp variable
            matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
            
    # Step 2: Reverse each row
    for i in range(n):
        # C-optimized reverse method is incredibly fast
        matrix[i].reverse()

if __name__ == "__main__":
    try:
        n = int(input("Enter the size of the matrix (n):\n").strip())
        
        matrix = []
        print("Enter the matrix rows (numbers separated by spaces):")
        for _ in range(n):
            row = list(map(int, input().strip().split()))
            if len(row) != n:
                print(f"Expected {n} elements, got {len(row)}.")
                exit()
            matrix.append(row)
            
        rotate(matrix)
        
        print("Rotated Matrix:")
        for row in matrix:
            print(row)
            
    except ValueError:
        print("Invalid input detected. Please ensure all inputs are valid integers.")